### PR TITLE
Replace storage db http client

### PR DESCRIPTION
**Pull request by guperrot**
Things to consider before you submit the PR:

* [ ] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [x] Did you add unit tests?
* [ ] Did you check UI tests on the sample app? They are not executed on CI.
* [ ] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?

## Description

Replace ingestion classes by http client in storage.

### CODE REVIEW

File: AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m [hunk 1]
```diff
@@ -66,15 +66,15 @@ @implementation MSDataStore
...
+    _tokenExchangeUrl = (NSURL *)[NSURL URLWithString:kMSDefaultApiUrl];
   }
   return self;
 }
 
 #pragma mark - Public
 
 + (void)setTokenExchangeUrl:(NSString *)tokenExchangeUrl {
-  [[MSDataStore sharedInstance] setTokenExchangeUrl:tokenExchangeUrl];
+  [[MSDataStore sharedInstance] setTokenExchangeUrl:(NSURL *)[NSURL URLWithString:tokenExchangeUrl]];
```
Comment (guperrot) [inline AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]: Can this throw if not a valid URL?

Comment (achocron) [reply AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]: nope

File: AppCenterDataStorage/AppCenterDataStorage/Internal/Client/MSTokenExchange.m [hunk 2]
```diff
@@ -20,32 +20,44 @@
...
   if (!cachedToken) {
 
-    // Payload.
+    // Serialize payload.
     NSError *jsonError;
     NSData *payloadData = [NSJSONSerialization dataWithJSONObject:@{kMSPartitions : @[ partition ]} options:0 error:&jsonError];
 
+    // Call token exchange service.
+    NSMutableDictionary *headers = [NSMutableDictionary new];
+    headers[kMSHeaderContentTypeKey] = kMSAppCenterContentType;
```
Comment (jaelim-ms) [inline AppCenterDataStorage/AppCenterDataStorage/Internal/Client/MSTokenExchange.m]: Not part of the change but could it be `kMSJsonContentType` instead?

File: AppCenterDataStorage/AppCenterDataStorage/Internal/Client/MSCosmosDb.m [hunk 3]
```diff
@@ -128,42 +128,41 @@ + (NSString *)documentUrlWithTokenResult:(MSTokenResult *)tokenResult documentId
...
-                                  completionHandler:(MSCosmosDbCompletionHandler)completionHandler {
-
+                                  completionHandler:(MSHttpRequestCompletionHandler)completionHandler {
   [MSCosmosDb performCosmosDbAsyncOperationWithHttpClient:httpClient
                                               tokenResult:tokenResult
                                                documentId:documentId
                                                httpMethod:httpMethod
                                                      body:body
                                         additionalHeaders:nil
+                                              offlineMode:NO
```
Comment (jaelim-ms) [inline AppCenterDataStorage/AppCenterDataStorage/Internal/Client/MSCosmosDb.m]: This shouldn't be `NO`. However the method is a dead code and only be used in test. Probably you can delete the method.

Comment (jaelim-ms) [reply AppCenterDataStorage/AppCenterDataStorage/Internal/Client/MSCosmosDb.m]: If you don't delete the method, you should add `offlineMode` to the method parameter.

File: AppCenterDataStorage/AppCenterDataStorageTests/MSDataStoreTests.m [hunk 4]
```diff
@@ -596,4 +648,121 @@ - (void)testSetOfflineMode {
...
+  OCMStub([self.cosmosDbMock performCosmosDbAsyncOperationWithHttpClient:OCMOCK_ANY
+                                                             tokenResult:OCMOCK_ANY
+                                                              documentId:OCMOCK_ANY
+                                                              httpMethod:OCMOCK_ANY
+                                                                    body:OCMOCK_ANY
+                                                       additionalHeaders:OCMOCK_ANY
+                                                             offlineMode:YES
+                                                       completionHandler:OCMOCK_ANY])
+      .andDo(^(__unused NSInvocation *invocation) {
+        calledWithOfflineModeEnabled = YES;
```
Comment (jaelim-ms) [inline AppCenterDataStorage/AppCenterDataStorageTests/MSDataStoreTests.m]: You can't do this with `OCMVerify`?

File: AppCenterDataStorage/AppCenterDataStorageTests/MSTokenExchangeTests.m [hunk 5]
```diff
@@ -71,21 +68,72 @@ - (void)testWhenNoCachedTokenNewTokenIsCached {
...
+                                            completionHandler:^(MSTokensResponse *tokenResponses, NSError *_Nullable returnError) {
+                                              XCTAssertNil(returnError);
+                                              returnedTokenResult = [tokenResponses tokens][0];
+                                              NSString *tokenValue = returnedTokenResult.token;
+                                              XCTAssertTrue([tokenValue isEqualToString:token]);
+                                              [completeExpectation fulfill];
+                                            }];
+
+  // Then
+  [self waitForExpectationsWithTimeout:5 handler:nil];
```
Comment (jaelim-ms) [inline AppCenterDataStorage/AppCenterDataStorageTests/MSTokenExchangeTests.m]: You don't need handler? We usually verify things within the handler.

File: AppCenterDataStorage/AppCenterDataStorageTests/MSTokenExchangeTests.m [hunk 6]
```diff
@@ -107,6 +157,9 @@ - (void)testWhenNoCachedTokenNewTokenIsCached {
   // Then
   [self waitForExpectationsWithTimeout:5 handler:nil];
```
Comment (jaelim-ms) [inline AppCenterDataStorage/AppCenterDataStorageTests/MSTokenExchangeTests.m]: Same here.

File: AppCenterDataStorage/AppCenterDataStorageTests/MSTokenExchangeTests.m [hunk 7]
```diff
@@ -136,7 +192,6 @@ - (void)testValidCachedTokenExists {
 
   // Then
   [self waitForExpectationsWithTimeout:5 handler:nil];
-  OCMVerifyAll(self.ingestionStoreMock);
```
Comment (jaelim-ms) [inline AppCenterDataStorage/AppCenterDataStorageTests/MSTokenExchangeTests.m]: Why do you delete this one? This is verifying `OCMReject` here, you need to call this one for `httpMock`.

File: AppCenterDataStorage/AppCenterDataStorage/Internal/Client/MSTokenExchange.m (hunk 2 above)
Comment (guperrot) [reply AppCenterDataStorage/AppCenterDataStorage/Internal/Client/MSTokenExchange.m]: Yes but out of scope of this PR.

File: AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m [hunk 8]
```diff
@@ -406,10 +414,8 @@ - (void)startWithChannelGroup:(id<MSChannelGroupProtocol>)channelGroup
       transmissionTargetToken:(nullable NSString *)token
               fromApplication:(BOOL)fromApplication {
   [super startWithChannelGroup:channelGroup appSecret:appSecret transmissionTargetToken:token fromApplication:fromApplication];
-
-  // Make sure that ingestion hasn't already been initialized.
```
[the remark this answers is not in the document]
Comment (guperrot) [reply AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]: this is for one collector start we don't need http client

File: AppCenterDataStorage/AppCenterDataStorage/Internal/Client/MSTokenExchange.h [hunk 9]
```diff
@@ -23,9 +23,11 @@ typedef void (^MSGetTokenAsyncCompletionHandler)(MSTokensResponse *tokenResponse
  * @param completionHandler callback that gets the token.
```
Comment (dhei) [inline AppCenterDataStorage/AppCenterDataStorage/Internal/Client/MSTokenExchange.h]: nit: update param

File: AppCenterDataStorage/AppCenterDataStorage/Internal/MSDataStoreInternal.h [hunk 10]
```diff
@@ -3,29 +3,24 @@
...
-#import "MSStorageIngestion.h"
 
 NS_ASSUME_NONNULL_BEGIN
 
+@protocol MSHttpClientProtocol;
+
 @interface MSDataStore <T : id <MSSerializableDocument>>() <MSServiceInternal>
 
 /**
  * An token exchange url that is used to get resouce tokens.
```
Comment (dhei) [inline AppCenterDataStorage/AppCenterDataStorage/Internal/MSDataStoreInternal.h]:
```suggestion
 * A token exchange url that is used to get resource tokens.
```

File: AppCenterDataStorage/AppCenterDataStorageTests/MSTokenExchangeTests.m (hunk 7 above)
Comment (guperrot) [reply AppCenterDataStorage/AppCenterDataStorageTests/MSTokenExchangeTests.m]: Please look at http://ocmock.org/reference/ and @clpolet  confirmed `OCMReject` does not need that.
You are probably confused with `OCMExpect`.

Comment (jaelim-ms) [reply AppCenterDataStorage/AppCenterDataStorageTests/MSTokenExchangeTests.m]:
```
Re-throwing fail fast exceptions in verify all
In fail-fast mode an exception might not cause the test to fail. This can happen when the call stack for the method does not end in the test. Fail fast exceptions will be re-thrown when OCMVerifyAll is called. This makes it possible to ensure that unwanted invocations from notifications etc. can be detected.
```

Comment (guperrot) [reply AppCenterDataStorage/AppCenterDataStorageTests/MSTokenExchangeTests.m]: I tested without and it rejected the call.
